### PR TITLE
[SPARK-31214][BUILD] Upgrade Janino to 3.1.2

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to upgrade Janino to 3.1.2 which is released recently.

Major changes were done for refactoring, as well as there're lots of "no commit message". Belows are the pairs of (commit title, commit) which seem to deal with some bugs or specific improvements (not purposed to refactor) after 3.0.15.

* Issue #119: Guarantee executing popOperand() in popUninitializedVariableOperand() via moving popOperand() out of "assert"
* Issue #116: replace operand to final target type if boxing conversion and widening reference conversion happen together
* Merged pull request `#114` "Grow the code for relocatables, and do fixup, and relocate".
  * https://github.com/janino-compiler/janino/commit/367c58e73e32014e5e1a5fc8225b7e5c17de00ad
* issue `#107`: Janino requires "org.codehaus.commons.compiler.io", but commons-compiler does not export this package
  * https://github.com/janino-compiler/janino/commit/f7d99596d43c56ab7b65cc8117db93f0fe4c6b6b
* Throw an NYI CompileException when a static interface method is invoked.
  * https://github.com/janino-compiler/janino/commit/efd3884983479f6e97adc2a270d02550276b7d3c
* Fixed the promotion of the array access index expression (see JLS7 15.13 Array Access Expressions)
  * https://github.com/janino-compiler/janino/commit/32fdb5f5f199c819394cae769e47d6fdcfa06b8f
* Issue `#104`: ClassLoaderIClassLoader 's ClassNotFoundException handle mechanism enhancement
  * https://github.com/janino-compiler/janino/commit/6e8a97d60908ecfa08b1d2d7c428a2631300516b

You can see the changelog from the link: http://janino-compiler.github.io/janino/changelog.html

### Why are the changes needed?

We got some report on failure on user's query which Janino throws error on compiling generated code. The issue is here: https://github.com/janino-compiler/janino/issues/113 It contains the information of generated code, symptom (error), and analysis of the bug, so please refer the link for more details.
Janino 3.1.1 contains the PR https://github.com/janino-compiler/janino/pull/114 which would enable Janino to succeed to compile user's query properly. I've also fixed a couple of more bugs as 3.1.1 made Spark UTs fail - hence we need to upgrade to 3.1.2.

Furthermore, from my testing, https://github.com/janino-compiler/janino/issues/90 (which Josh Rosen filed before) seems to be also resolved in 3.1.2 as well.

Looks like Janino is maintained by one person and there's no even version branches and releases/tags so we can't expect Janino maintainer to release a new bugfix version - hence have to try out new minor version.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing UTs.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Build failure seems to be related - just changed the PR to `WIP`. I'll kindly ask Janino maintainer to retain 3.0.x release but not sure it can be accepted. I'll look into details on the failures and try to deal with 3.1.1 as well.

**Comment by dongjoon-hyun** [comment]
Yes. The relevant UT failure seems to block us from upgrading.
```
failed to compile: org.codehaus.commons.compiler.InternalCompilerException: 
Compiling "GeneratedClass" in File 'generated.java', 
Line 1, Column 1: Compiling "MapEntries_0(InternalRow i)": 
Inconsistent operand stack: 
```

**Comment by dongjoon-hyun** [comment]
BTW, thank you for taking a look at this, @HeartSaVioR .

**Comment by HeartSaVioR** [comment]
UPDATE: I only have local dev. for now so cannot run all of Spark UTs easily, but I tried out some failed UTs with the custom build of Janino 3.0.15 + janino-compiler/janino#114 and they passed.

So it would be the ideal if we can have Janino 3.0.16 only adding janino-compiler/janino#114 on top of 3.0.15. I've asked it and waiting for the response. https://github.com/janino-compiler/janino/issues/115

In parallel, hopefully one of failing case is easy to reproduce - I've reported the issue via https://github.com/janino-compiler/janino/issues/116 . I'll try to cover more cases as there's a case having Janino 3.0.16 couldn't happen.

**Comment by HeartSaVioR** [comment]
FYI, while I'm still waiting for hearing the news from Janino, I also took the workaround into #27872 - while #27872 targets to master branch, I guess it would be more helpful for older versions, including 2.4.x. It would depend on how long we can wait; maybe we can simply apply #27872, and revert once Janino is ready.

**Comment by HeartSaVioR** [comment]
UPDATE: Somehow I fixed janino-compiler/janino#116 as well via janino-compiler/janino#117. Since I'm not sure this PR would fix all of UT failures we've seen before, I just made a tag in my repo and published via Jitpack.

Please note that I don't intend to maintain custom version Janino. Once we confirm it fixes the all test failures, I'll get back to Janino maintainer and ask for merging the PR & releasing next version. After then we can finally update the Janino version to the official one here.

**Comment by HeartSaVioR** [comment]
UPDATE: I also found janino-compiler/janino#119 and fixed it as well via janino-compiler/janino#120. Now Jenkins passes :)

If we feel OK to rely on UT passes as considering stability of Janino 3.1.x, we can go forward once Janino adopts these patches and do the next release.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by dongjoon-hyun** [comment]
Thank you for E2E experimenting. BTW, if there is a chance, let's test with JDK11 together.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
```
 org.apache.spark.deploy.yarn.LocalityPlacementStrategySuite.handle large number of containers and tasks (SPARK-18750)
 Error Details
org.scalatest.exceptions.TestFailedException: java.lang.StackOverflowError did not equal null
 Stack Trace
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: java.lang.StackOverflowError did not equal null
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:503)
	at org.apache.spark.deploy.yarn.LocalityPlacementStrategySuite.$anonfun$new$1(LocalityPlacementStrategySuite.scala:49)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:151)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:286)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:58)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
	at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:58)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:393)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:381)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:376)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:458)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1124)
	at org.scalatest.Suite.run$(Suite.scala:1106)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:518)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:58)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:58)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:317)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:510)
	at sbt.ForkMain$Run$2.call(ForkMain.java:296)
	at sbt.ForkMain$Run$2.call(ForkMain.java:286)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Doesn't seem to be related.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: For my own note, I checked https://github.com/janino-compiler/janino/commit/9e0c16c69c166f83541d63df70b42c9fbc278536

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: For the future reference:  I checked https://github.com/janino-compiler/janino/commit/9e0c16c69c166f83541d63df70b42c9fbc278536

**Review by HyukjinKwon — APPROVED** [review]
Logically looks good.

**Comment by HyukjinKwon** [comment]
cc @maropu, @kiszk, @rednaxelafx WDYT?

**Comment by HyukjinKwon** [comment]
cc @maropu, @kiszk, @rednaxelafx FYI.

**Comment by maropu** [comment]
Yea, merging this into the master looks fine to me if we have no performance regression (e.g., TPCDS) caused by the refactoring in janino.

**Comment by dongjoon-hyun** [comment]
In general, this looks okay to me for 3.1.0. BTW, @maropu ~ Do you have a plan to update your useful repo, https://github.com/maropu/spark-tpcds-datagen , for 3.0.0?

**Comment by kiszk** [comment]
From the technical perspective, I absolutely agree to merge this into Spark 3.0.

But, I have the mixed feeling in my mind. I agree with [this comment](https://github.com/apache/spark/pull/27872#issuecomment-601091320).   
It is a big change to upgrade the major version of Janino. Thus, I do not feel strong to merge this into 3.0 from the release management perspective. On the other hand, it is not good to stick the depreciated Janino 3.0 at Spark 3.0. From the maintenance perspective, I agree to merge this into Spark 3.0.

After a long thinking, I agree to merge this into Spark 3.0 if this PR does not cause performance regression. The Spark community appreciates it if Databricks team can measure the performance and run additional regression tests.  cc @gatorsmile

**Comment by maropu** [comment]
> In general, this looks okay to me for 3.1.0. BTW, @maropu ~ Do you have a plan to update your useful repo, https://github.com/maropu/spark-tpcds-datagen , for 3.0.0?

Ur, yea! I will. Thanks for letting me know!

**Comment by dongjoon-hyun** [comment]
Thank you, @maropu . The *benchmark* rang me your repository. :)

**Comment by gatorsmile** [comment]
@kiszk Yes. Let us do the perf regression tests before merging this. 

@xuanyuanking Could you run our internal benchmarking regression tests and see whether this upgrade will trigger the regressions?

**Comment by gatorsmile** [comment]
@HeartSaVioR @kiszk @maropu @rednaxelafx How confident are you about this new version especially when 3.1.1 are still not stable? 

Since Spark 3.0 is already code freeze, could you justify why we have to upgrade it?

**Comment by HyukjinKwon** [comment]
Personally I am a bit cautious whenever I see vendors are explicitly mentioned in OSS.

**Comment by HeartSaVioR** [comment]
So that's the matter of maintenance; I wouldn't expect we will get 3.0.17 easily, even though we fix the issue and provide the patch. The 3.0.x version line wasn't even maintained and even after branching it's defined as "deprecated".

Regarding confidence of stability, yeah, Spark is used to work with 3.0.x so 3.0.16 gives more confidence, though what I originally fixed in Janino is actually long-standing bug. There're still some known bugs in 3.0.x (SPARK-25987, and https://github.com/janino-compiler/janino/issues/90#issuecomment-509020465 Josh Rosen reported to Janino) which will never be fixed in 3.0.x I think, because these bugs don't exist in 3.1.x.

The question remains: have we ever constructed the way to justify the confidence of Janino stability? Any tests fail to catch the bugs I mentioned, so I don't think we have stability tests which verify edge-cases. In normal cases, a bunch of Spark UTs generate codes and Janino will be tested by these codes whenever we integrate Janino. If we evaluate the confidence from just test coverages, nothing looks to be changed.

I'd agree about general fear of adopting minor version which actually had major changes; that's why I asked about opinions in other PR. In the long run I think Janino 3.1.x is the way to go, but if adopting Janino 3.1.x becomes the burden for Spark 3.0.0, I'm OK to go with 3.0.16 in Spark 3.0 - any further Janino bugs may require Spark 3.1 to resolve.

**Comment by gatorsmile** [comment]
https://issues.apache.org/jira/browse/SPARK-25987 has been addressed in SPARK 3.0? 

Upgrading to Janino 3.0.16 looks much safer to me. Using the latest version is always risky to me. We normally avoid doing it for critical libraries we rely on. How about waiting for a few more months, after more projects try Janino 3.1.

**Comment by HeartSaVioR** [comment]
The code Spark generates are far from hand-written code which only few of users would have gone through the same path - the bugs we've figured out are edge-cases which only happen when we put lots of branch in a method, or the switch statement has bunch of statements (exceeding 32K of offset) in it.

If we consider Janino as a critical library we may need to consider it as something we need to investigate and collaborate, instead of treating it simply as a one of 3rd party.

I totally understand we may want to avoid the risk on Spark 3.0.0 as it's closer and closer to out the door, but if we ever fear about adopting Janino 3.1.x into Spark 3.1 and want to wait for some other one to encounter issues and deal with these instead of us, well... I'm not sure it will happen.

**Comment by kiszk** [comment]
@HyukjinKwon Sorry, I will take care of it. In my mind, there is a thought about whether [this great framework](https://databricks.com/session/a-framework-for-evaluating-the-performance-and-the-correctness-of-the-spark-sql-engine) can be used for this.

**Comment by dongjoon-hyun** [comment]
Hi, @HeartSaVioR .
The JIRA issue is changed into SPARK-31214.

**Comment by srowen** [comment]
I'm OK with janino 3.0.x now, but, yeah if 3.1.x is risky, it's not less risky in Spark 3.1, 3.2, etc. Maybe we simply cross that bridge when forced to.

**Comment by HeartSaVioR** [comment]
Thanks for sorting it out, @dongjoon-hyun !

Btw, I have been thinking that the concern was against Spark 3.0 instead of further minor versions, like Spark 3.1. e.g. for Spark 3.1, based on comments we seem to think OK to go with Janino 3.1.x.

I'll keep this open as basically I'm with @srowen - consider Janino 3.0.16 as the last version of 3.0.x, then we'll realize we should deal with Janino 3.1.x in the long run, IMHO in near future.

**Comment by dongjoon-hyun** [comment]
Hi, @HeartSaVioR . Could you rebase this PR once more?

**Comment by HeartSaVioR** [comment]
OK just rebased. Thanks for reminding.

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master for Apache Spark 3.1.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**
